feat: Adds nested NetworkObject support in prefabs

Clicky02 opened the pull request:
Allows users to spawn nested NetworkObjects in prefabs. This is done using a new dependent NetworkObjects system.

When you spawn a prefab with nested NetworkObjects, all of the nested NetworkObjects will also be spawned/synchronized. These nested NetworkObjects will be dependent on the root NetworkObject. This means they are initially synchronized in the same message as the root NetworkObject, and cannot outlive the root NetworkObject (when the root gets destroyed the depending NetworkObjects will be destroyed as well). Aside from these two exceptions, depending NetworkObjects behave identically to normal NetworkObjects.

To achieve this I made 3 major changes:
1. I added two hidden serialized fields to `NetworkObject` for saving its dependent NetworkObject and depending NetworkObjects. Respectively, these are `m_DependentNetworkObject ` and `m_DependingNetworkObjects`. They are determined in `OnValidate` and have public getters.
2. When `Spawn()` is called on a NetworkObject, it locally spawns depending NetworkObjects as well.
3. The `SceneObject` struct (which contains the data needed to spawn or initially synchronize NetworkObjects) now contains a list of `DependingSceneObject`s. This is a new struct that contains a subset of the properties of `SceneObject` along with a `IsSpawned` boolean.  When a SceneObject is spawned remotely, the information in the `DependingSceneObject` list is used to initialize the depending NetworkObjects. Any `DependingSceneObject` where `IsSpawned` is false has its corresponding NetworkObject destroyed instead of initialized. This occurs when a depending NetworkObject has been destroyed before the NetworkObject it is dependent on.

Resolves #2637

## Changelog

- Added: The ability to spawn (and properly network) prefabs with nested NetworkObjects

## Testing and Documentation

- Adds tests to ensure that:
    1. Player prefabs can have nested NetworkObjects
    2. Depending NetworkObjects are reparented properly
    4. Depending NetworkObjects are deleted properly
    5. Dependent NetworkObjects are deleted properly

Clicky02 committed:
feat(nested_network_objects): Added nested NetworkObject support

Introduces the concept of Dependent NetworkObjects. Dependent NetworkObjects cannot outlive the
NetworkObject they are dependent on. When spawning a prefab with nested NetworkObjects, all nested
NetworkObjects are dependent on the root NetworkObject. This means that nested NetworkObjects can
always be spawned/synchronized by spawning the root NetworkObject (and despawning any
NetworkObjects that are no longer spawned).

Resolves #2637

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=2645) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=2645) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/com.unity.netcode.gameobjects?pullRequest=2645) it.</sub>

Clicky02 committed:
update

Cleaned up code to be more specific, remove a log statement, hide serialized dependent
NetworkObject fields in editor, and correct typos.

Clicky02 committed:
update

Removed unnecessary serialization code

Clicky02 committed:
update

added public property to see what NetworkObjects are depending on a NetworkObject

Clicky02 committed:
fix

Fixed a bug throwing errors when spawning prefabs without children

Clicky02 committed:
fix

Fixed a bug where dependent network objects wouldn't be reparented to root on late-joining clients

Clicky02 committed:
fix

Fixed a bug where player prefabs could not have nested NetworkObjects

Clicky02 committed:
fix

Fixed a bug where despawned NetworkObjects would attempt to be serialized

Clicky02 committed:
test

Validates that nested network objects work properly.

Clicky02 committed:
changelog

Updated changelog to include nested NetworkObjects

Clicky02 commented:
PR has been updated to fix a number of issues and to add tests

razveck commented:
Any update on this? The PR is not merged

Toastarve commented:
Still waiting after almost a year :(
In my current production it would be rather necessary to have this feature...

SubatomicPlanets commented:
I need this too. This is the only feature that is making it impossible for me to use Netcode for GameObjects properly. Please make this happen!

edunad commented:
Any updates on merging this PR? Would make my life easier

Clicky02 commented:
@NoelStephensUnity Is this a feature that you/Unity would be interested in? I'd be more than happy to update my PR for the latest version if it is.

EmandM commented:
Hi all! Our team are currently making a push to address open pull requests. @Clicky02 if this is still something you're interested in landing, we can consider the feasibility of it.

To be transparent, from an initial pass over this PR this is touching some very high traffic areas and looks like a reasonably large change to the NetworkObject. As we now have the distributed authority network topology, we also have to consider the impacts of ownership changes, synchronization, and object distribution to this pattern. This change has a large surface area and so will need some very rigorous testing to be sure that all possible synchronization edge cases are covered.

Our current recommended approach to parenting is to have a top level NetworkObject and child object instead have NetworkBehaviours. Due to the design of Netcode for GameObjects, NetworkObjects are much more expensive over the network than NetworkBehaviours. We are currently working on integrating more tooling to make it easier to work with and re-parent child NetworkBehaviours.

if you are willing to work with us to design and test this feature, then we can go through the process of figuring out the next steps. As it stands, our team does not have the capacity to test this feature to the standards that are required for releasing new features into first-party packages.

EmandM commented:
Seeing as we haven't had a reply in a month, I'm going to close this PR as it is stale.
